src/VisualStudio/Core/Def/Telemetry/AggregatingTelemetryLogManager.cs

```diff
@@ -0,0 +1,100 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Internal.Log;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Microsoft.VisualStudio.Telemetry;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Telemetry
+{
+    /// <summary>
+    /// Manages creation and obtaining aggregated telemetry logs. Also, notifies logs to
+    /// send aggregated events every 30 minutes or upon disposal.
+    /// </summary>
+    internal sealed class AggregatingTelemetryLogManager : IDisposable
+    {
+        private readonly TelemetrySession _session;
+        private readonly Dictionary<FunctionId, AggregatingTelemetryLog> _aggregatingLogs;
+        private readonly object _lock;
+        private const int BatchedTelemetryCollectionPeriodInSeconds = 60 * 30;
+        private readonly AsyncBatchingWorkQueue _postTelemetryQueue;
+        private readonly CancellationTokenSource _cancellationTokenSource;
```

Comment:
surprsied this had it's own CTS, and isn't looking at IThreadingContext.DisposalToken.

Reply:
Was having issues trying to use IThreadingContext.DisposalToken per our offline conversation and reverted back to the way I originally coded this.